### PR TITLE
Add binary protocol/prepared statement support

### PR DESCRIPTION
**Pull request by brianmario**
As the title says, this adds support for the binary protocol, which is used by prepared statements.

There's still a little bit to do here, though mainly just wrapping up docs and some cleanup and adding a few more tests. But I wanted to get this open for review sooner than later.

I figured wrapping this with the Ruby gem/extension could happen in a separate pull request, but if it makes sense to just go ahead and add here let me know.

Some of the things I'd like to wrap up before calling this done:

- [x] Finish up writing client API docs
- [x] Finish up writing blocking API docs
- [x] Add tests for the blocking API
- [x] Fix null bitmap builder
- [x] Add more tests for edge cases
- [x] Document `trilogy_binary_value_t`, describing which C types map to which MySQL column types.

So happy to see this project finally made open source.

Thank you!

### CODE REVIEW

**Review by Mmkred — APPROVED** [review]

**Comment by brianmario** [comment]
@eileencodes hadn't seen this before, but after merging in your change from #5 it requires approval for the workflow to run on this pull. Make sense I guess. But just FYI for when one of you have a moment 🙏

**Comment by composerinteralia** [comment]
Approved. We've got the "Require approval for first-time contributors" setting selected, so we shouldn't need to do that for you again.

File: src/protocol.c
Comment (composerinteralia) [inline src/protocol.c:775]: I see something about this byte being used to indicate whether the type is unsigned. Is that something we care about here?

**Review by composerinteralia — COMMENTED** [review]
Looks like I need to approve the workflows every time you push to this branch, so I'll keep doing that. I think once this is merged they will run for you automatically on any future PRs.

I'm working my way through this, trying to make sure I understand everything as I go. Looks great so far!

File: src/protocol.c
Comment (brianmario) [reply src/protocol.c:775]: Totally missed coming back to this, good catch!

**Comment by brianmario** [comment]
Interesting test failure. Is there a way to re-run individual actions or does it require another push?

**Comment by composerinteralia** [comment]
I've seen that one before. I triggered a re-run

File: src/client.c
Comment (composerinteralia) [inline src/client.c:871]: nitpick for consistency, unless I am missing something:

```suggestion
    int err = begin_command_phase(&builder, conn, 0);
```

instead of `int err = 0;` above.

**Review by composerinteralia — COMMENTED** [review]
I looked through everything and it all seems to check out to me. 🎉 I'll leave it to somebody with more comfort in C than I to give a final review once you are happy with test coverage and such.

Adding this to Ruby gem makes sense to me as a separate PR, since this PR's already got plenty going on.

File: src/client.c
Comment (brianmario) [reply src/client.c:871]: Love it 👍

**Comment by brianmario** [comment]
@composerinteralia amazing thank you for giving it a look over! 🙏 

I'll see if I can find someone to help review the C-specific stuff when I'm ready for it.

**Comment by composerinteralia** [comment]
I played around with some blocking tests (modeled after your client tests) as part of my review. Feel free to use or throw away as necessary!

```c
TEST test_blocking_stmt_prepare()
{
    trilogy_conn_t conn;

    connect_conn(&conn);

    const char *query = "SELECT ?";
    trilogy_stmt_t stmt;

    int err = trilogy_stmt_prepare(&conn, query, strlen(query), &stmt);
    ASSERT_OK(err);

    ASSERT_EQ(1, stmt.parameter_count);

    trilogy_column_packet_t param;
    err = trilogy_read_full_column(&conn, &param);
    ASSERT_OK(err);

    ASSERT_EQ(1, stmt.column_count);

    trilogy_column_packet_t column_def;
    err = trilogy_read_full_column(&conn, &column_def);
    ASSERT_OK(err);

    trilogy_free(&conn);
    PASS();
}

TEST test_blocking_stmt_execute()
{
    trilogy_conn_t conn;

    connect_conn(&conn);

    const char *query = "SELECT ?";
    trilogy_stmt_t stmt;

    int err = trilogy_stmt_prepare(&conn, query, strlen(query), &stmt);
    ASSERT_OK(err);

    ASSERT_EQ(1, stmt.parameter_count);

    trilogy_column_packet_t param;
    err = trilogy_read_full_column(&conn, &param);
    ASSERT_OK(err);

    ASSERT_EQ(1, stmt.column_count);

    trilogy_column_packet_t column_def;
    err = trilogy_read_full_column(&conn, &column_def);
    ASSERT_OK(err);

    uint8_t flags = 0x00;
    const char str[] = {'t', 'e', 's', 't'};
    size_t len = sizeof(str);
    trilogy_binary_value_t binds[] = {
        {.is_null = false, .type = TRILOGY_TYPE_VAR_STRING, .as.str.data = str, .as.str.len = len}};

    uint64_t column_count;

    err = trilogy_stmt_execute(&conn, &stmt, flags, binds, &column_count);
    ASSERT_OK(err);

    ASSERT_EQ(1, column_count);

    trilogy_column_packet_t columns[1];
    err = trilogy_read_full_column(&conn, &columns[0]);
    ASSERT_OK(err);

    trilogy_binary_value_t values[1];
    err = trilogy_stmt_read_full_row(&conn, &stmt, columns, values);
    ASSERT_OK(err);

    ASSERT_MEM_EQ(values[0].as.str.data, "test", values[0].as.str.len);

    err = trilogy_stmt_read_full_row(&conn, &stmt, columns, values);
    ASSERT_EOF(err);

    trilogy_free(&conn);
    PASS();
}

TEST test_blocking_stmt_reset()
{
    trilogy_conn_t conn;

    connect_conn(&conn);

    const char *query = "SELECT ?";
    trilogy_stmt_t stmt;

    int err = trilogy_stmt_prepare(&conn, query, strlen(query), &stmt);
    ASSERT_OK(err);

    ASSERT_EQ(1, stmt.parameter_count);

    trilogy_column_packet_t param;
    err = trilogy_read_full_column(&conn, &param);
    ASSERT_OK(err);

    ASSERT_EQ(1, stmt.column_count);

    trilogy_column_packet_t column_def;
    err = trilogy_read_full_column(&conn, &column_def);
    ASSERT_OK(err);

    err = trilogy_stmt_reset(&conn, &stmt);
    ASSERT_OK(err);

    trilogy_free(&conn);
    PASS();
}

TEST test_blocking_stmt_close()
{
    trilogy_conn_t conn;

    connect_conn(&conn);

    const char *query = "SELECT ?";
    trilogy_stmt_t stmt;

    int err = trilogy_stmt_prepare(&conn, query, strlen(query), &stmt);
    ASSERT_OK(err);

    ASSERT_EQ(1, stmt.parameter_count);

    trilogy_column_packet_t param;
    err = trilogy_read_full_column(&conn, &param);
    ASSERT_OK(err);

    ASSERT_EQ(1, stmt.column_count);

    trilogy_column_packet_t column_def;
    err = trilogy_read_full_column(&conn, &column_def);
    ASSERT_OK(err);

    err = trilogy_stmt_close(&conn, &stmt);
    ASSERT_OK(err);

    trilogy_free(&conn);
    PASS();
}

```

**Comment by brianmario** [comment]
@composerinteralia Thanks! Dropped them in. I'll go back and add some more edge-case tests for everything to try and make sure we've covered some more surface area.

File: inc/trilogy/protocol.h
Comment (dbussink) [inline inc/trilogy/protocol.h:607]: Would it be useful to reference https://dev.mysql.com/doc/internals/en/binary-protocol-value.html here to explain where this comes from?

Comment (dbussink) [reply inc/trilogy/protocol.h:607]: Also might be good to annotation each part in the `union` then with which binary protocal value type(s) it implements?

File: inc/trilogy/protocol.h
Comment (dbussink) [inline inc/trilogy/protocol.h:550]: Adding a reference to https://dev.mysql.com/doc/internals/en/com-stmt-prepare-response.html might be useful here?

File: inc/trilogy/protocol.h
Comment (dbussink) [inline inc/trilogy/protocol.h:609]: I think with moving this to the end of the struct and then also moving `is_negative` to the end for date / date time might allow for tighter packing of the struct?

File: src/builder.c
Comment (dbussink) [inline src/builder.c:129]: Should the block above instead be in a header? If it's not, I don't think the define guards would be needed since `*.c` files should only be compiled once?

File: src/reader.c
Comment (dbussink) [inline src/reader.c:111]: As it's duplicated here, we might want to put it indeed into a header?

File: inc/trilogy/protocol.h
Comment (brianmario) [reply inc/trilogy/protocol.h:640]: Good call 👍

File: inc/trilogy/protocol.h
Comment (brianmario) [reply inc/trilogy/protocol.h:607]: Annotating the union is definitely a good idea 👍  It's on my internal list of things to do before taking this out of draft, I'll add it to the pull request body checklist so I don't forget ;)

In general, another documentation pass wouldn't hurt.

File: src/builder.c
Comment (brianmario) [reply src/builder.c:129]: To be honest, I kinda hated doing this 😅

But, these types are truly just helpers for the internal serialization implementation(s) and should never have to be used by a public user of this library. I know it's not a big deal to throw it in a header and just mark it private, but as we've both seen with `libmysqlclient` - whatever is exposed to the caller, they'll use and hack in ways we can't imagine.

If this stays in the C files, the guards are because the Ruby extension combines all the C files into a single source file for compilation. That process was actually something I wanted to have as a `Makefile` target for folks to more easily be able to embed this project in their own (which would combine everything into a single `.c`. and `.h`). I haven't checked if it exists yet, but it would be nice if it did.

That said, I could pretty easily be convinced to just put it in a header.

What do you think?

Comment (composerinteralia) [reply src/builder.c:129]: I think the guards were related to the Ruby library concatenating all the source files: https://github.com/github/trilogy/blob/main/contrib/ruby/ext/trilogy-ruby/extconf.rb#L3-L6

Comment (composerinteralia) [reply src/builder.c:129]: Ha! You beat me to this comment by a minute

Comment (dbussink) [reply src/builder.c:129]: @brianmario Would it alternatively work to do this inline? For example:

```c
int trilogy_reader_get_double(trilogy_reader_t *reader, double *out)
    union {double d; uint64_t u;} double_val;
    double_val.d = val;

    int rc = trilogy_reader_get_uint64(reader, &double_val.u);
    if (rc != TRILOGY_OK) {
        return rc;
    }

    *out = double_val.d;
    return TRILOGY_OK;
}
```

That way there's no separate type definition necessary and there's no issue with combining C files either?

Comment (brianmario) [reply src/builder.c:129]: Ah yeah I like that 👍

File: test/blocking_test.c
Comment (dbussink) [inline test/blocking_test.c:484]: Would it be possible to create one test for each type that's supported?

Comment (brianmario) [reply test/blocking_test.c:484]: Yeah that's the plan. I just haven't had a chance to sit down to do it yet 😅

**Comment by adrianna-chang-shopify** [comment]
👋 Hi folks -- I'm gonna pick this up and try to get it across the finish line. @brianmario I can either start a new PR with commits cherry-picked from yours, or if you give me write access to your fork I can push updates here.

Do either you or @composerinteralia remember what sorts of "edge cases" we were still looking to test? (I've yet to take a look at the code in depth, so it's not immediately obvious to me what gaps are left to fill). I'll take a look at the test suite for mysql2's prepared statements for inspiration as well, but figured I'd see if either of y'all could remember where this PR was at 😂

**Comment by brianmario** [comment]
Hi!

@adrianna-chang-shopify I can totally give you access to my fork, not a problem at all.

In terms of where it was left off, it _should_ be ready to go other than my wanting there to be some more tests. I haven't had time to get back to writing them myself, but the rest of the code should be ready to go.

**Comment by composerinteralia** [comment]
I recall this looking fairly ready when I reviewed it a while back. I'd be fine to merging this as-is after a rebase or whatever and then we can fill in whatever additional test coverage we want (e.g. testing with all the different types) separately. Some of that test coverage could probably also be done via the Ruby client once we've added support there.

**Comment by adrianna-chang-shopify** [comment]
Okay, so I've added prepared statement tests for all of the MySQL types, two of which are still failing and have me a bit stumped:
```
Expected: TRILOGY_OK
     Got: TRILOGY_ERR
F
FAIL test_blocking_stmt_execute_year: (TRILOGY_OK) != ((err)) (test/blocking_test.c:758)
```

```
FAIL test_blocking_stmt_execute_float: values[0].as.flt != float_val (test/blocking_test.c:420)
```

The first one I can't manage to repro locally, which is making it difficult to debug. (I apologize for all the noise with the builds, I was trying to debug on CI 😅 ). Seems like building the statement itself is problematic -- if anyone wants to take a stab at reproing locally that would be great 🙏 

The second one _does_ fail locally, but the failure is bizarre. Basically, when Trilogy reads the column back from the resultset, it thinks that the type is `TRILOGY_TYPE_DOUBLE` instead of `TRILOGY_TYPE_FLOAT`. This is shown if print out the value as a float vs a double:
```diff
diff --git a/test/blocking_test.c b/test/blocking_test.c
index 083a132..6da78c0 100644
--- a/test/blocking_test.c
+++ b/test/blocking_test.c
@@ -411,12 +411,16 @@ TEST test_blocking_stmt_execute_float() {
 
     trilogy_column_packet_t columns[1];
     err = trilogy_read_full_column(&conn, &columns[0]);
+    printf("Column type: %d\n", columns[0].type);
     ASSERT_OK(err);
 
     trilogy_binary_value_t values[1];
     err = trilogy_stmt_read_full_row(&conn, &stmt, columns, values);
     ASSERT_OK(err);
 
+    printf("Value as float: %f\n", values[0].as.flt); // wrong
+    printf("Values as double: %lf\n", values[0].as.dbl);
+
     ASSERT_EQ(values[0].as.flt, float_val);
 
     err = trilogy_stmt_read_full_row(&conn, &stmt, columns, values);
```

```
Column type: 5 # TRILOGY_TYPE_DOUBLE
Value as float: 0.000000
Values as double: 1234.500000
F
FAIL test_blocking_stmt_execute_float: values[0].as.flt != float_val (test/blocking_test.c:424)
```

That value is definitely being written correctly:
```diff
diff --git a/src/protocol.c b/src/protocol.c
index 232bb97..21a0577 100644
--- a/src/protocol.c
+++ b/src/protocol.c
@@ -781,6 +781,7 @@ int trilogy_build_stmt_execute_packet(trilogy_builder_t *builder, uint32_t stmt_
         CHECKED(trilogy_builder_write_uint8(builder, 0x1));
 
         for (i = 0; i < num_binds; i++) {
+            printf("Writing type: %d\n", binds[i].type);
             CHECKED(trilogy_builder_write_uint8(builder, binds[i].type));
 
             if (binds[i].is_unsigned) {
```

```
Writing type: 4
Column type: 5 # TRILOGY_TYPE_DOUBLE
Value as float: 0.000000
Values as double: 1234.500000
F
FAIL test_blocking_stmt_execute_float: values[0].as.flt != float_val (test/blocking_test.c:424)
```

But when it's read in `#trilogy_parse_column_packet`, we think it's a double:
```diff
diff --git a/src/protocol.c b/src/protocol.c
index 232bb97..d207ba6 100644
--- a/src/protocol.c
+++ b/src/protocol.c
@@ -379,6 +379,7 @@ int trilogy_parse_column_packet(const uint8_t *buff, size_t len, bool field_list
 
     uint8_t type;
     CHECKED(trilogy_reader_get_uint8(&reader, &type));
+    printf("Read column type: %d\n", type);
     out_packet->type = type;
 
     CHECKED(trilogy_reader_get_uint16(&reader, &out_packet->flags));
```
```
Writing type: 4
Read column type: 5
Column type: 5
Value as float: 0.000000
Values as double: 1234.500000
F
FAIL test_blocking_stmt_execute_float: values[0].as.flt != float_val (test/blocking_test.c:424)
```

Would love input if either of you have it @brianmario @composerinteralia! Maybe I'm missing something simple.

**Comment by matthewd** [comment]
> ```
> Value as float: 0.000000
> Value as double: 1234.500000
> ```

It sounds like it _is_ a double [as opposed to being a float that's mis-tagged]. Perhaps the server accepts incoming floats for compatibility, but just upcasts them to a double immediately?

Ooh, maybe relevant:

> Using FLOAT might give you some unexpected problems because all calculations in MariaDB are done with double precision.

**Comment by adrianna-chang-shopify** [comment]
Oh, good find @matthewd ! I'll dig a bit more -- but yeah, if that's the behaviour on the server side, there's not much we can do there 😄

**Comment by adrianna-chang-shopify** [comment]
Okay, this is finally ready for review 🙈 I'll rebase off main and squash the commits once we deem it merge-able.

**Comment by brianmario** [comment]
Thanks for adding the rest of these tests! 🙏

It looks good to me, but I'll wait for someone with merge access to do a review as well 😉

**Review by composerinteralia — APPROVED** [review]

**Comment by brianmario** [comment]
🎉

**Pull request merged**
